libpromises/generic_agent.c

```diff
@@ -116,6 +138,236 @@ void MarkAsPolicyServer(EvalContext *ctx)
     Log(LOG_LEVEL_VERBOSE, "Additional class defined: policy_server");
 }
 
+bool CheckContextOrClassmatch(EvalContext *ctx, const char* c)
+{
+    ClassTableIterator *iter = EvalContextClassTableIteratorNewGlobal(ctx, NULL, true, true);
+    StringSet *global_matches = ClassesMatching(ctx, iter, c, NULL, true); // returns early
+
+    bool found = (StringSetSize(global_matches) > 0);
+
+    StringSetDestroy(global_matches);
+    ClassTableIteratorDestroy(iter);
+
+    if (found)
+    {
+        return found;
+    }
+
+    // does it look like a regex? It's not a class expression then
+    // (these characters are invalid in class expressions and will
+    // give errors)
+    if (strchr(c, '*') ||
+        strchr(c, '+') ||
+        strchr(c, '['))
+    {
+        return false;
+    }
+
+    return IsDefinedClass(ctx, c);
+}
+
+void LoadAugmentsFiles(EvalContext *ctx, const char* filename)
+{
+    Buffer *filebuf = BufferNewFrom(filename, strlen(filename));
+    Buffer *expbuf = BufferNew();
+    ExpandScalar(ctx, NULL, "this", BufferData(filebuf), expbuf);
+    if (strstr(BufferData(expbuf), "/.json"))
+    {
+        Log(LOG_LEVEL_DEBUG, "Skipping augments file '%s' because it failed to expand the base filename, resulting in '%s'",
+            BufferData(filebuf),
+            BufferData(expbuf));
+    }
```

Comment:
This function nests incredibly deeply, and is hard to read because of code running off to the right. Many of them are due to code paths not taken if there is an error, and I think this is one of the cases where it is justified to use ``goto`` in order to quickly escape from the error condition and down to the cleanup. I suggest adding a ``goto`` to this block and then letting the next block continue outside of the ``else`` block.

This goes for several of the error conditions in the file; most of them can have their ``else`` block replaced by ``goto`` down to the end. Just make sure you goto to the correct cleanup location.

Here is an example: https://github.com/cfengine/core/blob/master/libpromises/locks.c#L954-L961. A lot of ``else`` blocks are avoided that way.